keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,553 @@
+
+## Motivation
+
+Kubernetes clusters are increasingly deployed in multi-zone environments.
+Network traffic is routed randomly to any endpoint matching a Service. Some
+users might want the traffic to stay in the same zone for the following
+reasons:
+  - Cost savings: Keeping traffic within a zone can limit cross-zone networking
+costs.
+  - Performance: Traffic within a zone is often more performant than traffic
+leaving the zone.
+
+In this KEP we are going to focus on avoiding cross-zone traffic when in-zone
+endpoints would suffice. We're attempting to provide a simple and more
+automatic approach to topology aware routing. This API will still allow users
+to require traffic to stay within a zone, or simply to indicate that they
+prefer to keep traffic in the same zone if there's sufficient capacity. With
+this approach no new API fields are required, we can accomplish this simply
+with a new annotation on Service.
+
+
+### Goals
+- Provide a simple way for users to indicate their preference for keeping
+traffic in zone.
+- Provide different approaches to users who have a strong desire over the
+routing policy:
+  - Balanced: Use the original approach to ensure traffic is distributed across
+  all endpoints for a Service.
+  - RequireZone: Only route to endpoints within the same zone.
+  - PreferZone: Prefer routing to endpoints within the same zone when they are
+  available.
+- Use EndpointSlice subsetting as the primary mechanism for topology aware
+routing.
+  - Improve kube-proxy scalability by limiting the number of endpoints it needs
+  to process.
+- Minimize churn of EndpointSlices while doing topology aware distribution.
+- Minimize the number of new EndpointSlices required.
+- Provide a simple API that does not require any new API fields.
+
+### Non-Goals
+- Real-time distribution rebalancing based on traffic load or distribution
+feedback or metrics.
+- Providing a way for kube-proxy to consume subsets of EndpointSlices (this is
+a separate KEP).
+- Providing a way to require Node local traffic ([this is a separate KEP](https://github.com/kubernetes/enhancements/pull/1944)).
+- Multi-cluster topology aware routing (this same pattern may be useful there
+though).
+- Region based topology aware routing (this may come later).
+- Ensuring that endpoints are distributed evenly across zones.
+
+## Proposal
+
+### Terminology
+
+- **EndpointSlice Group**: One or multiple EndpointSlices that will be consumed
+by the same set of zones.
+- **Endpoints**: In this KEP, “endpoints” are used to describe individual Pod
+IPs. These are not references to Kubernetes Endpoints resources. We also omit
+the unready pods that number of endpoints = number of pods.
+
+### Configuration
+
+This proposal builds off of the [EndpointSlice API](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/).
+
+We propose a new Service annotation: `endpointslice.kubernetes.io/same-zone`
```

Comment:
heh naming party :smile:

Reply:
Haha, this will be tricky name to get right. I like the idea of zone-affinity, what would the values be here? I'm not sure `Prefer` or `Require` would make as much sense with this name.

Reply:
what about `hard` , `soft` and `none` ?

this aligns with node affinity terminoloy https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity

Reply:
One of our goals with this kind of naming pattern is that we could use the same approach for node and region (`same-node`, `same-region`). Unfortunately I think that would be difficult/confusing for an `affinity` based name as `node-affinity` already has a different meaning in k8s. Still very interested in naming suggestions here, just hesitant to use something that might not work as well for node or region.

Reply:
Finding the appropriate name here is definitely challenging. We actually explored options similar to what you're suggesting here. My first suggestion for an annotation was `endpointslice.kubernetes.io/slicing-policy` which seems similar to the concept of an endpoint selection policy. We ended up moving away from that since it seemed too closely tied to implementation and would not translate well for node-local config.

We also explored algorithm/policy names and created a [doc with possible combinations](https://docs.google.com/document/d/1Iul7rRTx_knezvPdwqZ9SQxEYUUXMTDnjCeYmNmcEGA/edit?usp=sharing). In this case it felt like there were simply too many potential combinations.

As I've thought about this KEP more, I think the title/scope may actually be a bit too broad here. This is fundamentally about trying to find smarter ways to subset EndpointSlices so that:
- Consumers only need to process a fraction of endpoints (significant scalability win).
- Consumers only get the endpoints closest to them.

This does not feel like it is the final solution for topology aware routing, just a step in the right direction. This will still allow more advanced logic to be built into consumers that could potentially make adjustments based on feedback/metrics. If we take that kind of a more limited approach to this KEP, maybe a narrower name/annotation would also make sense?